File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h10-w/low_level/lrcu_state.cpp
// lrcu_state.cpp

#include <iostream>
#include <thread>
#include <chrono>

#include "common/common_def.h"
#include "common/h10w_def.h"
#include "h10-w/h10w_low_level.h"

using namespace humanoid;
using namespace humanoid::h10w;

int main() {
  std::cout << "=== LRCU State Reading Test ===" << std::endl;

  H10wLowLevel low_level;
  LRCUState lrcu_state;

  // Read LRCU state for 10 seconds
  constexpr int DURATION_MS = 10000;
  constexpr int READ_INTERVAL_MS = 500;
  constexpr int read_count = DURATION_MS / READ_INTERVAL_MS;

  for (int i = 0; i < read_count; ++i) {
    auto start = std::chrono::high_resolution_clock::now();

    // Read LRCU state
    int ret = low_level.readLrcuState(lrcu_state);
    if (ret != 0) {
      std::cout << "Failed to read LRCU state!" << std::endl;
      return -1;
    }

    // Print LRCU state information
    std::cout << "=== LRCU State Reading " << i + 1 << " ===" << std::endl;
    std::cout << "Mode: " << static_cast<int>(lrcu_state.mode) << std::endl;
    std::cout << "LED Mode: " << static_cast<int>(lrcu_state.led_mode)
              << std::endl;
    std::cout << "LED Frequency: " << static_cast<int>(lrcu_state.led_frequency)
              << std::endl;
    std::cout << "LED Color - R: " << static_cast<int>(lrcu_state.led_r)
              << ", G: " << static_cast<int>(lrcu_state.led_g)
              << ", B: " << static_cast<int>(lrcu_state.led_b) << std::endl;
    std::cout << std::endl;

    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed =
        std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::this_thread::sleep_for(std::chrono::milliseconds(READ_INTERVAL_MS) -
                                elapsed);
  }

  std::cout << "LRCU state reading test completed successfully" << std::endl;
  return 0;
}

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h10-w/low_level/lrcu_state.py
#!/usr/bin/env python3

import humanoid_sdk_py
from humanoid_sdk_py import h10w
import time


def main():
    print("=== LRCU State Reading Test ===")

    # Initialize low level object
    low_level = h10w.H10wLowLevel()

    # Read LRCU state for 10 seconds
    DURATION_MS = 10000
    READ_INTERVAL_MS = 500
    read_count = DURATION_MS // READ_INTERVAL_MS

    for i in range(read_count):
        start_time = time.time()

        # Read LRCU state
        ret, lrcu_state = low_level.readLrcuState()
        if ret != 0:
            print("Failed to read LRCU state!")
            return -1

        # Print LRCU state information
        print(f"=== LRCU State Reading {i + 1} ===")
        print(f"Mode: {lrcu_state.mode}")
        print(f"LED Mode: {lrcu_state.led_mode}")
        print(f"LED Frequency: {lrcu_state.led_frequency}")
        print(f"LED Color - R: {lrcu_state.led_r}, G: {lrcu_state.led_g}, B: {lrcu_state.led_b}")
        print()

        elapsed_time = (time.time() - start_time) * 1000
        sleep_time = READ_INTERVAL_MS - elapsed_time
        if sleep_time > 0:
            time.sleep(sleep_time / 1000.0)

    print("LRCU state reading test completed successfully")
    return 0


if __name__ == "__main__":
    exit(main())
